python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+        self._next_chunk_index = 0
+        self._chunk_object_ids = []
+        self._ready_indices = []
+        self._ready_objects = []
+        if not hasattr(iterable, "__len__"):
+            iterable = [iterable]
+        self._iterator = iter(iterable)
+        if chunksize:
+            self._chunksize = chunksize
+        else:
+            self._chunksize = pool._calculate_chunksize(iterable)
+        self._total_chunks = int(math.ceil(len(iterable) / chunksize))
+        self._result_thread = ResultThread(
+            [], total_object_ids=self._total_chunks)
+        self._result_thread.start()
+
+        for _ in range(len(pool._actor_pool)):
+            self._submit_next_chunk()
+
+    def _submit_next_chunk(self):
+        if len(self._chunk_object_ids) >= self._total_chunks:
+            return
+
+        actor_index = len(self._chunk_object_ids) % len(self._pool._actor_pool)
+        new_chunk_id = self._pool._submit_chunk(self._func, self._iterator,
+                                                self._chunksize, actor_index)
+        self._chunk_object_ids.append(new_chunk_id)
+        self._ready_indices.append(None)
+        self._result_thread.add_object_id(new_chunk_id)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+
+class OrderedIMapIterator(IMapIterator):
+    def next(self, timeout=None):
+        if len(self._ready_objects) != 0:
+            return self._ready_objects.pop(0)
+
+        if self._next_chunk_index == len(self._chunk_object_ids):
+            raise StopIteration
+
+        while timeout is None or timeout > 0:
+            start = time.time()
+            try:
+                index = self._result_thread.next_ready_index(timeout=timeout)
+                self._submit_next_chunk()
+            except queue.Empty:
+                raise TimeoutError
+            self._ready_indices[index] = True
+            if index == self._next_chunk_index:
+                break
+            if timeout is not None:
+                timeout = max(0, timeout - (time.time() - start))
+
+        while self._next_chunk_index < len(
+                self._chunk_object_ids
+        ) and self._ready_indices[self._next_chunk_index]:
+            self._ready_objects.extend(
+                self._result_thread.results()[self._next_chunk_index])
+            self._next_chunk_index += 1
+
+        return self._ready_objects.pop(0)
+
+
+class UnorderedIMapIterator(IMapIterator):
+    def next(self, timeout=None):
+        if len(self._ready_objects) != 0:
+            return self._ready_objects.pop(0)
+
+        if self._next_chunk_index == len(self._chunk_object_ids):
+            raise StopIteration
+
+        try:
+            index = self._result_thread.next_ready_index(timeout=timeout)
+            self._submit_next_chunk()
+        except queue.Empty:
+            raise TimeoutError
+
+        self._ready_objects.extend(self._result_thread.results()[index])
```

Comment:
`self._result_thread.results()[index]` looks like it's making an an expensive call and throwing away most of its return value. Is this the case? If so, I'd pass in `index` so it can return only the relevant portion. Same with other similar instances of this in the PR.